pinot-plugins/pinot-file-system/pinot-adls/src/main/java/org/apache/pinot/plugin/filesystem/AzureGen2PinotFS.java

```diff
@@ -0,0 +1,447 @@
+  private static final String FILE_SYSTEM_NAME = "fileSystemName";
+
+  private static final String HTTPS_URL_PREFIX = "https://";
+  private static final String DIRECTORY_DELIMITER = "/";
+
+  private static final String AZURE_STORAGE_DNS_SUFFIX = ".dfs.core.windows.net";
+  private static final String AZURE_BLOB_DNS_SUFFIX = ".blob.core.windows.net";
+  private static final String PATH_ALREADY_EXISTS_ERROR_CODE = "PathAlreadyExists";
+  private static final String IS_DIRECTORY_KEY = "hdi_isfolder";
+
+  private static final int NOT_FOUND_STATUS_CODE = 404;
+  private static final int ALREADY_EXISTS_STATUS_CODE = 409;
+
+  // Azure Data Lake Gen2's block size is 4MB
+  private static final int BUFFER_SIZE = 4 * 1024 * 1024;
+
+  private DataLakeFileSystemClient _fileSystemClient;
+  private BlobServiceClient _blobServiceClient;
+
+  @Override
+  public void init(Configuration config) {
+    // Azure storage account name
+    String accountName = config.getString(ACCOUNT_NAME);
+    String accessKey = config.getString(ACCESS_KEY);
+    String fileSystemName = config.getString(FILE_SYSTEM_NAME);
+    String dfsServiceEndpointUrl = HTTPS_URL_PREFIX + accountName + AZURE_STORAGE_DNS_SUFFIX;
+    String blobServiceEndpointUrl = HTTPS_URL_PREFIX + accountName + AZURE_BLOB_DNS_SUFFIX;
+
+    StorageSharedKeyCredential sharedKeyCredential = new StorageSharedKeyCredential(accountName, accessKey);
+
+    DataLakeServiceClient serviceClient = new DataLakeServiceClientBuilder().credential(sharedKeyCredential)
+        .endpoint(dfsServiceEndpointUrl)
+        .buildClient();
+
+    _blobServiceClient =
+        new BlobServiceClientBuilder().credential(sharedKeyCredential).endpoint(blobServiceEndpointUrl).buildClient();
+    _fileSystemClient = serviceClient.getFileSystemClient(fileSystemName);
+    LOGGER.info("AzureGen2PinotFS is initialized (accountName={}, fileSystemName={}, dfsServiceEndpointUrl={}, "
+        + "blobServiceEndpointUrl={})", accountName, fileSystemName, dfsServiceEndpointUrl, blobServiceEndpointUrl);
+  }
+
+  @Override
+  public boolean mkdir(URI uri) throws IOException {
+    LOGGER.info("mkdir is called with uri='{}'", uri);
+    try {
+      // By default, create directory call will overwrite if the path already exists. Setting IfNoneMatch = "*" to
+      // prevent overwrite. https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/create
+      DataLakeRequestConditions requestConditions = new DataLakeRequestConditions().setIfNoneMatch("*");
+      _fileSystemClient.createDirectoryWithResponse(convertUriToAzureStylePath(uri), null, null, null, null,
+          requestConditions, null, null);
+      return true;
+    } catch (DataLakeStorageException e) {
+      // If the path already exists, doing nothing and return true
+      if (e.getStatusCode() == ALREADY_EXISTS_STATUS_CODE && e.getErrorCode().equals(PATH_ALREADY_EXISTS_ERROR_CODE)) {
+        return true;
```

Comment:
I'm explicitly following `pinotFS` interface.

```
  /**
   * Creates a new directory. If parent directories are not created, it will create them.
   * If the directory exists, it will return true without doing anything.
   * @return true if mkdir is successful
   * @throws IOException on IO failure
   */
```